designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,349 @@
+2. AllowedDisruptionByReason = minNodeCount[reason] - unhealthyNodes - totalDisruptingNodes
+3. minimumDisruptionAllowed = min(minNodeCount[reason], minNodeCount[default]) - unhealthyNodes - totalDisruptingNodes
+
+### Scenarios
+We will go through scenarios that users are expecting to face to drive home if the api can satisfy the users requirements for the end users of these features
+- Block Drift, but allow Expiration
+- Fully Block Drift && Expiration, but allow consolidation to occur 
+- Block Drift && Expiration, but allow emptiness 
+- Limit Drift && Expiration while allowing aggressive but not unbounded consolidation
+
+Note that these samples are written with the intention to think through our design, and more formal documentation will be prepared for common customer scenarios.
+
+#### Block Drift and allow Expiration 
+Customers who roll out aggressive node image upgrades don't want to roll the nodes due to drift, but want to use node expiration to force users onto compliant node images eventually. The scenario here is an ML Training workload. Ideally we want to let it bake as long as possible, but not at the cost of compromising the security of the cluster. 
+
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenEmpty 
+    expireAfter: 7d 
+    budgets:
+    - nodes: 0
+      reasons: [drifted] 
+    - nodes: "100%" // All Disruption Other than drift(Expiration and Empty Nodes) are allowed
+```
+
+We could also allow a semantic like this
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenEmpty 
+    expireAfter: 7d 
+    budgets:
+    - nodes: "100%" # Always allow Expiration and Emptiness to disrupt 
+      reasons: [expired, empty]
+    - nodes: 0
+```
+So in this case, we would take the minimum node count, and it would block the 100% expiration budget of expired and Empty. 
+But this semantic would not work for equation 3. Equation three will take the minimum between minNodeCount[reason] and the minNodeCount[default/unspecifed].  
+
+#### Fully Blocking Drift && Expiration while allowing consolidation
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized 
+    expireAfter: Never 
+    budgets:
+    - nodes: 0
+      reasons: [drifted, expired] 
+    - nodes: "100%" // All Disruption Other than drift and expiration are allowed
+```
+
+#### Fully Blocking Drift and Expiration while allowing emptiness 
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenEmpty
+    expireAfter: Never 
```

Comment:
The interaction of `expireAfter: Never` and `budgets: [nodes: 0, reasons: [expired]]` is a bit unsatisfying. I wonder if there are edge cases that'll emerge from having two ways to accomplish the same result

Reply:
I agree. I used the expireAfter: Never controls here to call that out implicitly that we have multiple controls for blocking expire. The never semantic is conflicting with the budgets semantic. 

I can define 

```
disruption: 
     expireAfter: Never 
     
     budgets: 
        - nodes: 10 
          reasons: ["expired"] 
          
 ```
 
In this conflicting example, our behavior still adheres to the principle we set in the budgets. In this case we respect the lower disruption setting anyway. If i set expire never we will not expire nodes. 

I can similarly define a budget with conflicting values.  
```
disruption: 
    budgets: 
      - nodes: 10 
         reasons: ["expired"]
      - nodes: 15 
         reasons: ["expired"]
```

We take 10 rather than 15. In that sense expireAfter is disrupt 0 due to expire and the budget of 10 from the disruption budgets, we still take the disruption action that is least disruptive.